lib/mqtt/mqtt.cc

```diff
@@ -235,6 +226,58 @@ namespace
 		return -ETIMEDOUT;
 	}
 
+	/**
+	 * Note from the MQTT 3.1.1 spec:
+	 * The Server MUST allow ClientIds which are between 1 and 23 UTF-8 encoded
+	 * bytes in length, and that contain only the characters
+	 * "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	 *
+	 * Note from us: UTF-8 encoding of 0-9, a-z, A-Z, is 1 Byte per
+	 * character.
+	 */
+	constexpr const char clientIDCharacters[] =
+	  "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+	constexpr const int maxClientIDLength = 23;
+
+	/**
+	 * Helper to check if a client ID is valid according to the MQTT
+	 * specification.
+	 */
+	bool is_valid_client_id(const char *id, size_t length)
+	{
+		if (length > maxClientIDLength || !length)
+		{
+			return false;
+		}
+
+		for (size_t i = 0; i < length; i++)
+		{
+			if (id[i] == '\0' || strchr(clientIDCharacters, id[i]) == nullptr)
+			{
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Helper to check if a buffer contains a \0 character.
+	 */
+	bool has_null_character(const char *buffer, size_t length)
+	{
+		// TODO use `memchr` (or a string view and `find`) when
+		// available
+		for (size_t i = 0; i < length; i++)
+		{
+			if (buffer[i] == '\0')
+			{
+				return true;
+			}
+		}
+		return false;
+	}
```

Comment:
```suggestion
bool has_null_character(std::string_view buffer)
{
	return buffer.find('\0') != std::string_view::npos;
}
```

C++23 makes this simpler, but our stdlib is a bit too old.

Reply:
That's what I initially had, but it doesn't link:
```
error: ld.lld: error: undefined symbol: memchr
>>> referenced by debug.hh:588 (/mnt/workspace/cheriot-rtos/sdk/include/debug.hh:588)
>>>               build/cheriot/cheriot/release/MQTT.compartment:()
```
We don't have `memchr`, hence the comment.

Reply:
We should probably just add it, it’s very small.

Reply:
Can do. But how about we merge this PR first, I add `memchr` ASAP, and go back to this code once it's in? I don't want to delay the merge of this PR by a week if possible